app/websites/routes.py

```diff
@@ -133,31 +69,65 @@ async def search_websites(
 ) -> SearchResponse:
     logger = structlog.get_logger("websites.search")
 
-    cursor_obj: Optional[SearchWebsiteCursor] = None
-    if cursor is not None:
-        match SearchWebsiteCursor.from_base64(cursor):
-            case Ok(obj):
-                cursor_obj = obj
-            case Err(err):
-                logger.warning(
-                    "Invalid cursor received",
-                    cursor=cursor,
-                    error=err,
-                )
-                raise HTTPException(
-                    status_code=status.HTTP_400_BAD_REQUEST,
-                    detail=[{"loc": ["query", "cursor"], **err.model_dump()}],
-                )
+    match cursor:
+        case None:
+            cursor_obj = SearchWebsiteCursor()
+        case c if (parsed := SearchWebsiteCursor.from_base64(c)).is_ok():
+            cursor_obj = parsed.unwrap()
+        case _:
+            logger.warning("Invalid cursor received", cursor=cursor)
+            raise HTTPException(
+                status_code=status.HTTP_400_BAD_REQUEST,
+                detail=[
+                    {
+                        "loc": ["query", "cursor"],
+                        "msg": "invalid key-value pair format",
+                        "type": "invalid_cursor_format",
+                    },
+                ],
+            )
+
+    searched = await search.search_websites(db, q or "", cursor_obj, limit)
+
+    grouped = groupby(
+        searched.entries,
+        key=lambda e: f"{e.campus_id}${e.department_id}${e.office_id}",
+    )
 
-    logger.info("Cursor object", cursor_obj=cursor_obj)
+    result: list[SearchResultEntry] = []
+    for group_key, search_entries in grouped:
+        search_entries = list(search_entries)
+        if len(search_entries) == 0:
+            continue
+
+        first = search_entries[0]
+        websites = [
+            SearchResultWebsite(
+                id=entry.website_id,
+                name=entry.website_name,
+                url=Url(entry.website_url),
+            )
+            for entry in search_entries
+        ]
+        result.append(
+            SearchResultEntry(
+                id=group_key,
+                campus=first.campus_name,
+                department=first.department_name,
+                office=first.office_name,
+                websites=websites,
+            )
+        )
 
     return SearchResponse(
-        result=[],
+        result=result,
         pagination=Pagination(
-            next_cursor="",
-            has_next=False,
-            num_results=0,
-            total_results=0,
+            next_cursor=searched.next_cursor.to_base64()
+            if searched.next_cursor is not None
+            else None,
+            has_next=searched.next_cursor is not None,
```

Comment:
`has_next` cannot be extracted, since the type system cannot infer`next_cursor is not None` from `has_next == True`